### DISCUSSION TITLE
Can I use pandoc in a lua filter to render parts of the AST to the current target format?

### DISCUSSION THREAD

**Discussion by retorquere** [jgm/pandoc#10054] [Q&A]
I have a lua filter that converts markdown to docx/odt Zotero live citations. I do this by replacing Cite elements with RawInline docx/odt in the filter. Parts of the citation, such as citationSuffix, may have inline markup (Quoted, Underline etc) which I'd have to convert to RawInline myself. Is there a way to use the pandoc itself in the lua filter to render the citationSuffix to the raw XML for the target format? Perhaps by (ab)using pandoc.utils.citeproc?

**Comment by denismaier** [comment]
Maybe pandoc.write ?

Gesendet von Outlook für Android<https://aka.ms/AAb9ysg>
________________________________
From: Emiliano Heyns ***@***.***>
Sent: Sunday, August 4, 2024 4:03:24 PM
To: jgm/pandoc ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [jgm/pandoc] Can I use pandoc in a lua filter to render parts of the AST to the current target format? (Discussion #10054)


I have a lua filter that converts markdown to docx/odt Zotero live citations. I do this by replacing Cite elements with RawInline docx/odt in the filter. Parts of the citation, such as citationSuffix, may have inline markup (Quoted, Underline etc) which I'd have to convert to RawInline myself. Is there a way to use the pandoc itself in the lua filter to render the citationSuffix to the raw XML for the target format? Perhaps by (ab)using pandoc.utils.citeproc?

**Comment by badumont** [comment]
You can let Pandoc handle the prefix and the suffix itself by returning a List made up of the Inlines of the prefix and suffix and of the citation ID formatted by you. Here is a proof of concept:

function Cite(cite)
  local res = pandoc.List()
  for _, citation in ipairs(cite.citations) do
    if citation.prefix then
      res:extend(citation.prefix)
      res:insert(pandoc.Space())
    end
    res:insert(pandoc.RawInline('opendocument', '<stg>' .. citation.id .. '</stg>'))
    if citation.suffix then
      res:insert(pandoc.Space())
      res:extend(citation.suffix)
    end
  end
  return res
end

Of course, <stg> should be replaced by the correct formatting. The trailing space in the prefix and the leading space in the suffix are removed by Pandoc when parsing: the filter adds them, but it should not do this unconditionally as I did here (e.g. not when the suffix begins with a comma).

As the citation suffix is not interpreted by citeproc, the locator cannot be wrapped in braces or they will be printed as is.

**Reply by retorquere** [reply]
Great, that should work nicely!

I just realized I do have an additional problem though - parts of the citation can have formatting, and I need to put that in the raw inline as html-in-JSON-in-XML. I don't suppose I can convince pandoc to render some things to html even though the output format is docx? Although the html subset supported is fairly limited - perhaps easy enough to do that by hand.

Can I have the Cite handler do it's own walk of the subnodes, with behavior specific to nodes within a Cite node?

**Comment by badumont** [comment]
Could you please provide an example? I am not sure that I understand what you wish/have to do.

**Reply by retorquere** [reply]
I'm trying to turn this:

```
[@kooijAltenTextzeugenJesajabuches1981, 113. He argues, "Man kann folglich die Vorlage der LXX Jes nur mit der [Vorlage]{.underline} von Q^a^ vergleichen" ("Consequently, one can only compare the *Vorlage* of LXX with the [*Vorlage*]{.underline} of Q^a^"\; author's translation)]
```

into

```
<w:r>
  <w:instrText xml:space="preserve"> ADDIN ZOTERO_ITEM CSL_CITATION {
    "properties": {
      "formattedCitation":"...",
      "noteIndex": 0,
      "unsorted":false
    },
    "citationItems":[
      {                                                                                                                                                                          "locator": "113",
        "id": 105,
        "label": "page",
        "suffix": ". He argues, “Man kann folglich die Vorlage der LXX Jes nur mit der &lt;u&gt;Vorlage&lt;/u&gt; von Q&lt;sup&gt;a&lt;/sup&gt; vergleichen” (“Consequently, one can only compare the &lt;b&gt;Vorlage&lt;/b&gt; of LXX with the &lt;u&gt;&lt;b&gt;Vorlage&lt;/b&gt;&lt;/u&gt; of Q&lt;sup&gt;a&lt;/sup&gt;”; author’s translation)",
        "uris":["http://zotero.org/users/8230813/items/ZCHECLES"],
        "itemData": {
          "citation-key":"kooijAltenTextzeugenJesajabuches1981",
          "title":"book",
          "type":"book",
          "id":
          "kooijAltenTextzeugenJesajabuches1981"
        },
        "prefix": ""
      }
    ],
    "schema":"https://github.com/citation-style-language/schema/raw/master/csl-citation.json",
    "citationID":"00000001"
}   </w:instrText>
</w:r>
```

essentially replicating what docx+citations does, but in a way that connects the document to the users Zotero library.

**Comment by badumont** [comment]
Thank you, I realize that my solution will not work since the whole citation, including the prefix and the suffix, will have to be in the JSON object included in <w:instrText>. As you suggest, you will probably have to pass the prefix and suffix Inlines to a filter that will convert every element to a string containing the desired markup, using the walk method of the Inlines type. I don't think that using pandoc.write would be of great help here, for you would have to modify the output heavily (e.g. converting "strong" to "b", "<" to "&lt;", etc.).

**Reply by retorquere** [reply]
And there's no way to make pandoc output `<b>` instead of `<strong>`?

**Reply by retorquere** [reply]
This seems to work pretty well:

```
function stringify(node)
  local doc = pandoc.Pandoc({ node })
  return pandoc.write(doc, 'html')
    :gsub('<em>', '<i>')
    :gsub('</em>', '</i>')
    :gsub('<strong>', '<b>')
    :gsub('</strong>', '</b>')
    :gsub('[<>&]', { ['&'] = '&amp;', ['<'] = '&lt;', ['>'] = '&gt;' })
end
```

**Comment by badumont** [comment]
It seems to be more compact than what I envisioned :-) You need to remove the <p> tag as well though.
